votor/src/consensus_pool.rs

```diff
@@ -192,40 +158,24 @@ impl ConsensusPool {
             let (limit, vote_types) = certificate_limits_and_vote_types(&cert_type);
             let accumulated_stake = vote_types
                 .iter()
-                .filter_map(|vote_type| {
-                    Some(match self.vote_pools.get(&(slot, *vote_type))? {
-                        VotePool::SimpleVotePool(pool) => pool.total_stake(),
-                        VotePool::DuplicateBlockVotePool(pool) => {
-                            pool.total_stake_by_block_id(block_id.as_ref().expect(
-                                "Duplicate block pool for {vote_type:?} expects a block id for \
-                                 certificate {cert_type:?}",
-                            ))
-                        }
-                    })
+                .map(|vote_type| {
+                    self.vote_pools
+                        .get(&slot)
+                        .map_or(0, |p| p.get_stake(vote_type, vote.block_id()))
                 })
                 .sum::<Stake>();
+
             if accumulated_stake as f64 / (total_stake as f64) < limit {
                 continue;
             }
             let mut cert_builder = CertificateBuilder::new(cert_type);
-            vote_types.iter().for_each(|vote_type| {
-                if let Some(vote_pool) = self.vote_pools.get(&(slot, *vote_type)) {
-                    match vote_pool {
-                        VotePool::SimpleVotePool(pool) => {
-                            cert_builder.aggregate(pool.votes()).unwrap();
-                        }
-                        VotePool::DuplicateBlockVotePool(pool) => {
-                            let block_id = block_id.as_ref().expect(
-                                "Duplicate block pool for {vote_type:?} expects a block id for \
-                                 certificate {cert_type:?}",
-                            );
-                            if let Some(votes) = pool.votes(block_id) {
-                                cert_builder.aggregate(votes).unwrap();
-                            }
-                        }
-                    };
+            for vote_type in vote_types {
+                if let Some(vote_pool) = self.vote_pools.get(&slot) {
+                    cert_builder
+                        .aggregate(&vote_pool.get_votes(vote_type, vote.block_id()))
```

Comment:
Just saying: we originally designed the `aggregate()` interface to avoid a copy every time.

Reply:
I agree that this is a drawback of this PR.  The PR has added a TODO to fix this issue.  I already have a branch where I am fixing it and I will make the proposal shortly.